reset.c

```diff
@@ -0,0 +1,888 @@
+            printf("Failed to set output data (%d)\n", errno);
+            break;
+        }
+
+        memcpy(bam_get_aux(outdata), bam_get_aux(bamdata), bam_get_l_aux(bamdata));
+        outdata->l_data += bam_get_l_aux(bamdata);  //TODO htslib doesnt have an interface for this?
+
+        errno = 0;
+        //write bam data to output
+        if (0 > (ret_w = sam_write1(outfile, out_samhdr, outdata)))
+        {
+            printf("Failed to write data (%d)\n", errno);
+            break;
+        }
+        // wrote the data, continue read/write cycle
+        errno = 0;
+    }
+
+    if (-1 > ret_r || 0 > ret_w)
+    {   //some error
+        printf("Error during %s (%d)!\n", (-1 > ret_r)? "read" : "write", errno);
+        ret_r = EXIT_FAILURE;
+    }
+    else // no error!
+        ret_r = EXIT_SUCCESS;
+
+    // clean up and return result
+    sam_hdr_destroy(in_samhdr);
+    sam_hdr_destroy(out_samhdr);
+    bam_destroy1(bamdata);
+    bam_destroy1(outdata);
+    return ret_r;
+}
+
+/// cleanup - free up allocations made
+/** @param config - pointer to internal configuration data
+returns nothing
+*/
+void cleanup(conf_data *config)
+{
+    khiter_t iter = 0;
+    const char *tmp = NULL;
+    kh_destroy(aux_set, config->auxtagsin);
+    for (iter = kh_begin(config->pgout); iter != kh_end(config->pgout); ++iter) {
+        tmp = (kh_exist(config->pgout,iter)) ? kh_key(config->pgout, iter) : NULL;
+        if (tmp)
+            free((void*)tmp);
+        tmp = NULL;
+    }
+    kh_destroy(pg_set, config->pgout);
+}
+
+// reset main, return 0 on success and error code on failure
+// deals with a single file
+/// main_reset - starts the reset of data
+/** @param argc - count of arguments
+ *  @param argv - pointer to array of arguments
+returns 1 on failure 0 on success
+*/
+int main_reset(int argc, char *argv[])
+{
+    static const struct option lopts[] = {
+        SAM_OPT_GLOBAL_OPTIONS('-', '-', 0, '-', '-', 0),       //let output format and thread count be given by user - long options
+        {"no-RG", no_argument, NULL, 1},                        //no RG lines in output, default is to keep them
+        //reject PG lines from input, default is to keep them (i.e. option not given); without optional filename, all PGs removed and those given in file are filtered when optional filename is given
+        {"reject-PG", optional_argument, NULL, 'p'},
+        {NULL, 0, NULL, 0}
+    };
+    samFile *infile = NULL, *outfile = NULL;
+    sam_global_args ga = SAM_GLOBAL_ARGS_INIT;
+    htsThreadPool tpool = {NULL, 0};
+    const char *inname = NULL, *outname = NULL, *auxkeepname = NULL, *auxrejname = NULL, *pgrejname = NULL;
+    int c = 0, ret = 0;
+    char outmode[4] = "w", *args = NULL;
+    conf_data resetconf = {1, 1, NULL, NULL, 0};                //keep RGs and PGs by default
+
+    if (argc < 5)
+    {
+        //not enogh args
+        usage(stdout);
+        return EXIT_FAILURE;
+    }
```

Comment:
This has a number of problems.

Firstly, when using no arguments (`samtools reset`) it's almost always because the user is explicitly asking to get the usage statement, so that's fine to go to stdout, but if we *need* 5 arguments for correct usage and they specified 3 then that's a usage error and it should report to stderr.

However in this case I think it's incorrect anyway as you have a hard requirement on `-i file -o file`.  Generally we don't use `-i file` and just let `file` be the last thing on the command line.  Eg see `samtools sort` usage statement.

The use of `-o file` is good (thank you), but generally it should be optional.  As should the input filename.  When they're not specified the command should read and write to stdin / stderr, assuming it's not doing random access or needing multiple passes over the input stream.  This enables it to work cleanly on unix pipes.

Despite that, `-o file` is our preference over `> file` as it avoids the potential for mixing errors with the output file (especially when combined with the `nohup` command).

Reply:
-i is removed and last parameter is considered as input. if -o is not given, "-" will be used for stdout.